modules/apm/src/main/java/org/elasticsearch/tracing/apm/APMTracer.java

```diff
@@ -0,0 +1,394 @@
+     * exists to make it possible to use scopes in the few situation where it makes sense.
+     *
+     * @param spanId the ID of a currently-open span for which to open a scope.
+     * @return a method to close the scope when you are finished with it.
+     */
+    @Override
+    public Releasable withScope(String spanId) {
+        final Context context = spans.get(spanId);
+        if (context != null) {
+            var scope = context.makeCurrent();
+            return scope::close;
+        }
+        return () -> {};
+    }
+
+    private void setSpanAttributes(ThreadContext threadContext, @Nullable Map<String, Object> spanAttributes, SpanBuilder spanBuilder) {
+        if (spanAttributes != null) {
+            for (Map.Entry<String, Object> entry : spanAttributes.entrySet()) {
+                final String key = entry.getKey();
+                final Object value = entry.getValue();
+                if (value instanceof String) {
+                    spanBuilder.setAttribute(key, (String) value);
+                } else if (value instanceof Long) {
+                    spanBuilder.setAttribute(key, (Long) value);
+                } else if (value instanceof Integer) {
+                    spanBuilder.setAttribute(key, (Integer) value);
+                } else if (value instanceof Double) {
+                    spanBuilder.setAttribute(key, (Double) value);
+                } else if (value instanceof Boolean) {
+                    spanBuilder.setAttribute(key, (Boolean) value);
+                } else {
+                    throw new IllegalArgumentException(
+                        "span attributes do not support value type of [" + value.getClass().getCanonicalName() + "]"
+                    );
+                }
+            }
+
+            final boolean isHttpSpan = spanAttributes.keySet().stream().anyMatch(key -> key.startsWith("http."));
+            spanBuilder.setSpanKind(isHttpSpan ? SpanKind.SERVER : SpanKind.INTERNAL);
+        } else {
+            spanBuilder.setSpanKind(SpanKind.INTERNAL);
+        }
+
+        spanBuilder.setAttribute(org.elasticsearch.tracing.Tracer.AttributeKeys.NODE_NAME, nodeName);
+        spanBuilder.setAttribute(org.elasticsearch.tracing.Tracer.AttributeKeys.CLUSTER_NAME, clusterName);
+
+        final String xOpaqueId = threadContext.getHeader(Task.X_OPAQUE_ID_HTTP_HEADER);
+        if (xOpaqueId != null) {
+            spanBuilder.setAttribute("es.x-opaque-id", xOpaqueId);
+        }
+    }
+
+    @Override
+    public void addError(String spanId, Throwable throwable) {
+        final var span = Span.fromContextOrNull(spans.get(spanId));
+        if (span != null) {
+            span.recordException(throwable);
+        }
+    }
+
+    @Override
+    public void setAttribute(String spanId, String key, boolean value) {
+        final var span = Span.fromContextOrNull(spans.get(spanId));
+        if (span != null) {
+            span.setAttribute(key, value);
```

Comment:
How does the `Span` work? Is this a blocking IO call? Or is it non-blocking, in which case what happens if the other side is slow and we start queuing up work here?

Reply:
It's implementation-dependent, but it absolutely shouldn't be blocked. The APM Java agent buffers spans and send them on a separate thread. Spans will get dropped if there's slowdown on the APM server. From the [docs FAQ](https://www.elastic.co/guide/en/apm/agent/java/master/faq.html):

> It’s important to note that the latency of an application won’t be affected much by the agent (in the order of single-digit microseconds), even if you sample at 100%. However, the background reporter thread has some work to do for serializing and gzipping events. If your application is not CPU bound, this shouldn’t matter much. Note that if the APM Server can’t handle all the events, the agent will drop data to not crash your application. It will then also not serialize and gzip the events.

We'll ship with a sample rate of 20% by default, possibly lower after we've tried it out with internal clusters.